Rb/simplified_hvdc

rbolgaryn opened the pull request.

jd-lara commented on src/common.jl:
Why does this matter. The DCBuses are for MultiTerminal HVDC

rbolgaryn replied:
We get an error otherwise

jd-lara replied:
Why? and what's the error these checks for types usually give me the ick

GabrielKS reviewed (commented):
If the objection is just one of code style, there's the multiple dispatch alternative:
```julia
_get_bustype_wrapper(::PSY.Bus) = missing  # since `nothing` is a possible option for `ACBus.bustype`
_get_bustype_wrapper(bus::PSY.ACBus) = PSY.get_bustype(bus)

bt = PSY.get_bustype(bus)
ismissing(bt) && continue
```

But I agree in general that we're not super great at thoughtfully handling the `DCBus` case across PowerFlows (maybe across Sienna?). We should probably solve that problem more holistically.

jd-lara commented:
> If the objection is just one of code style, there's the multiple dispatch alternative:
> 
> ```julia
> _get_bustype_wrapper(::PSY.Bus) = missing  # since `nothing` is a possible option for `ACBus.bustype`
> _get_bustype_wrapper(bus::PSY.ACBus) = PSY.get_bustype(bus)
> 
> bt = PSY.get_bustype(bus)
> ismissing(bt) && continue
> ```
> 
> But I agree in general that we're not super great at thoughtfully handling the `DCBus` case across PowerFlows (maybe across Sienna?). We should probably solve that problem more holistically.

Why would we ever get a DCBus in this context? It is an impossibility for this approach, there has to be something more fundamentally wrong.

rbolgaryn commented:
This was the case giving me the issues:

```Julia
sys = build_system(PSISystems, "sys10_pjm_ac_dc")
julia> data = PowerFlowData(ACPowerFlow(), sys, check_connectivity=false)


ERROR: MethodError: no method matching get_bustype(::PowerSystems.DCBus)
The function `get_bustype` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  get_bustype(::PowerSystems.ACBus)
   @ PowerSystems ~/repos/PowerSystems.jl/src/models/generated/ACBus.jl:108

Stacktrace:
 [1] _initialize_bus_data!(bus_type::Vector{…}, bus_angles::Vector{…}, bus_magnitude::Vector{…}, temp_bus_map::Dict{…}, bus_lookup::Dict{…}, sys::PowerSystems.System)
   @ PowerFlows ~/repos/PowerFlows.jl/src/common.jl:122
 [2] make_powerflowdata(sys::PowerSystems.System, time_steps::Int64, power_network_matrix::PowerNetworkMatrices.Ybus{…}, aux_network_matrix::Nothing, n_buses::Int64, n_branches::Int64, bus_lookup::Dict{…}, branch_lookup::Dict{…}, temp_bus_map::Dict{…}, branch_type::Vector{…}, timestep_map::Dict{…}, valid_ix::Vector{…}, neighbors::Vector{…}, converged::Vector{…}, loss_factors::Nothing, calculate_loss_factors::Bool)
   @ PowerFlows ~/repos/PowerFlows.jl/src/common.jl:234
 [3] PowerFlowData(pf::ACPowerFlow{NewtonRaphsonACPowerFlow}, sys::PowerSystems.System; time_steps::Int64, timestep_names::Vector{String}, check_connectivity::Bool)
   @ PowerFlows ~/repos/PowerFlows.jl/src/PowerFlowData.jl:243
 [4] top-level scope
   @ REPL[10]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

rbolgaryn commented:
So you can see how the DC bus is there and causing the initialization of the PowerFlowData to crash

jd-lara commented:
> So you can see how the DC bus is there and causing the initialization of the PowerFlowData to crash. In the test case in PSI I ended up using instead, with TwoTerminalGenericHVDCLine, it does not occur. Nevertheless, it crashing when DC buses are present is still a bug.

The correct fix is to change the places where we cal `get_bus` to ACBus instead of Bus... this is come legacy code before we added DCBuses. 

<img width="499" alt="image" src="https://github.com/user-attachments/assets/082c60f8-77d9-4da4-89a4-bdc451442751" />

rbolgaryn commented:
> > So you can see how the DC bus is there and causing the initialization of the PowerFlowData to crash. In the test case in PSI I ended up using instead, with TwoTerminalGenericHVDCLine, it does not occur. Nevertheless, it crashing when DC buses are present is still a bug.
> 
> The correct fix is to change the places where we cal `get_bus` to ACBus instead of Bus... this is come legacy code before we added DCBuses.
> 
> <img alt="image" width="499" src="https://private-user-images.githubusercontent.com/16385323/438321751-082c60f8-77d9-4da4-89a4-bdc451442751.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU4NzIyMDUsIm5iZiI6MTc0NTg3MTkwNSwicGF0aCI6Ii8xNjM4NTMyMy80MzgzMjE3NTEtMDgyYzYwZjgtNzdkOS00ZGE0LTg5YTQtYmRjNDUxNDQyNzUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI4VDIwMjUwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxZGU4YjU5ZmQ3NTJjOTY1YzRlOTFjMDg1NGFhZjU3MGEyMGI2ZTEyOGIzNDM3YzU4OGU4ZTU5NTA1ZmExYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dD6mg4sf4BUgvrMpJl95Z9sH6RMug0nVeB9Wr3mJTWA">

I removed all the mentions of DCBus

GabrielKS reviewed (approved):
Works for me if we're sure this all will only ever need to work with `ACBus`

jd-lara commented:
> Works for me if we're sure this all will only ever need to work with `ACBus`

Yes, I mean there is no AC PowerFlow within DC Buses
